File: frontend/lib/monteCarlo.ts
export type MCResult = {
  price: number;
  stderr: number;
  ciLow: number;
  ciHigh: number;
};

// Mulberry32 — fast seedable PRNG
function makePrng(seed: number) {
  let s = seed >>> 0;
  return () => {
    s |= 0; s = s + 0x6d2b79f5 | 0;
    let t = Math.imul(s ^ s >>> 15, 1 | s);
    t = t + Math.imul(t ^ t >>> 7, 61 | t) ^ t;
    return ((t ^ t >>> 14) >>> 0) / 4294967296;
  };
}

// Box-Muller: uniform pair → standard normal
function boxMuller(u1: number, u2: number): number {
  return Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
}

export function mcPrice(
  S: number, K: number, T: number, r: number, sigma: number,
  optionType: "call" | "put",
  nPaths: number = 50_000,
  seed: number = 42,
): MCResult {
  const rand = makePrng(seed);
  const drift = (r - 0.5 * sigma * sigma) * T;
  const vol = sigma * Math.sqrt(T);
  const disc = Math.exp(-r * T);

  const payoffs = new Float64Array(nPaths);

  for (let i = 0; i < nPaths; i++) {
    const Z = boxMuller(rand(), rand());
    const Sp = S * Math.exp(drift + vol * Z);
    const Sm = S * Math.exp(drift - vol * Z);
    const pp = optionType === "call" ? Math.max(Sp - K, 0) : Math.max(K - Sp, 0);
    const pm = optionType === "call" ? Math.max(Sm - K, 0) : Math.max(K - Sm, 0);
    payoffs[i] = disc * (pp + pm) / 2;
  }

  let sum = 0, sum2 = 0;
  for (let i = 0; i < nPaths; i++) { sum += payoffs[i]; sum2 += payoffs[i] ** 2; }
  const mean = sum / nPaths;
  const variance = sum2 / nPaths - mean ** 2;
  const stderr = Math.sqrt(variance / nPaths);

  return { price: mean, stderr, ciLow: mean - 1.96 * stderr, ciHigh: mean + 1.96 * stderr };
}

File: src/monte_carlo.py
"""Monte Carlo pricer — European options with antithetic variance reduction."""

import numpy as np
from dataclasses import dataclass


@dataclass
class MCResult:
    price: float
    stderr: float       # standard error of the estimate
    ci_low: float       # 95% confidence interval lower bound
    ci_high: float      # 95% confidence interval upper bound
    n_paths: int


def mc_price(
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    option_type: str = "call",
    n_paths: int = 100_000,
    seed: int | None = 42,
) -> MCResult:
    """
    Price a European option by Monte Carlo simulation.

    Uses antithetic variates: for each draw Z we also simulate -Z,
    which halves variance at virtually no extra cost.

    Parameters
    ----------
    S           : spot price
    K           : strike
    T           : time to maturity in years
    r           : continuous risk-free rate
    sigma       : annualised volatility
    option_type : "call" or "put"
    n_paths     : number of simulation paths (antithetic → effective 2×n_paths)
    seed        : random seed for reproducibility
    """
    if option_type not in ("call", "put"):
        raise ValueError(f"option_type must be 'call' or 'put', got '{option_type}'")

    rng = np.random.default_rng(seed)

    # Draw n_paths standard normals
    Z = rng.standard_normal(n_paths)

    # Itô-corrected log-return: (r - σ²/2)T + σ√T·Z
    log_return = (r - 0.5 * sigma**2) * T + sigma * np.sqrt(T) * Z

    # Antithetic: simulate both Z and -Z
    S_plus  = S * np.exp(log_return)
    S_minus = S * np.exp((r - 0.5 * sigma**2) * T - sigma * np.sqrt(T) * Z)

    # Payoffs
    if option_type == "call":
        payoffs = 0.5 * (np.maximum(S_plus - K, 0) + np.maximum(S_minus - K, 0))
    else:
        payoffs = 0.5 * (np.maximum(K - S_plus, 0) + np.maximum(K - S_minus, 0))

    # Discount to present value
    discount = np.exp(-r * T)
    pv = discount * payoffs

    price  = float(pv.mean())
    stderr = float(pv.std(ddof=1) / np.sqrt(n_paths))

    return MCResult(
        price=price,
        stderr=stderr,
        ci_low=price  - 1.96 * stderr,
        ci_high=price + 1.96 * stderr,
        n_paths=n_paths,
    )
